[PERSISTENCE] SavePoints and Rollbacks design document (Issue #493)

deblasis opened the pull request:
## Description

This PR adds the `SAVEPOINTS_ROLLBACKS.md` document that represents a design proposal and an implementation guideline.

## Issue

Fixes #493 

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [x] Documentation
- [ ] Other <!-- add details here if it a different type of change -->

## List of changes

- Added design document

## Testing

N/A 

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

Olshansk commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+Performing operations like: copying datadirs, compressing them, etc. is probably not the best approach.
+Having a first-class support for savepoints and rollbacks would be, IMHO, a much better solution.
+
+A more advanced/flexible version of **Savepoints** could be to efficiently serialize the state into an artifact (non-trivial also because we have multiple data-stores but definitely possible), maybe leveraging previous savepoints to only store the changes that have been made since the last one and/or using some version of a WAL (Write-Ahead log that records the changes that happened).
```
1. Snapshotting the key-value store (i.e. serialising it) is also an idea that crossed my mind but as much as I like to avoid premature optimisation, I felt like it's asking for trouble.

2. Regarding the "WAL". What if we have a "WBL" with columns `key prev_height new_height prev_value new_value` that could be stored on disk and enable only updating/reverting the modified keys?

Olshansk commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+- The action of creating a savepoint must be easy to perform (e.g. a CLI command and/or a flag being passed to the node binary when starting it)
+- The action of creating a savepoint must be as fast as possible
+- The operator should be informed with meaningful messages about the progress of the savepoint creation process (telemetry, logging and stdout come to mind)
+- It should be, as much as possible, compact
```
What do you mean by compact?

deblasis replied:
I mean packed together, dense, zipped with a good compression ratio since speed of dissemination (download/upload) is IMHO more important than the one-off speed of compression.

Should I clarify?

Olshansk commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+- The action of creating a savepoint must be as fast as possible
+- The operator should be informed with meaningful messages about the progress of the savepoint creation process (telemetry, logging and stdout come to mind)
+- It should be, as much as possible, compact
+- It must have some form of integrity check mechanism (e.g. checksum/hash verification and maybe even a signature that could be very useful in the case of a social rollback)
```
The blocks contains quorum certs (i.e. the signatures) as well as the state hash. By definition, it's all verified & signed.

deblasis replied:
What I mean is that the archive itself should be verifiable as well so that if it's not signed you don't even bother opening it.

Zero-trust.

Olshansk replied:
Remove maybe - set the requirements to what they should be

Olshansk commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+
+### Further improvements
+
+- Savepoints could be disseminated/retrieved using other networks like Tor/IPFS/etc., this would free-up bandwidth on the main network and could be used in conjunction with `FastSync` to speed up the process of bootstrapping a new node without overwhelming the `P2P` network with a lot of traffic that is not `Protocol`-related. This could be very important when the network reaches critical mass in ...
```
I see, you really do look at savepoints as snapshots. 

Do you see the solution for a node-specific savepoint (i.e. I updated one tree but not another and crashed) and a network wide (i.e. a data dir snapshot) as having the same solution?

deblasis replied:
I think it can very much build on top of what I have in mind.

Similarly to what databases do, we'd need some `data integrity check` at startup that verifies that all the datastores are correct.

It can very much rely on the statehash I guess...

How?

KISS, back of the envelope implementation:
- retrieve the latest valid statehash (from file, from `SuperValidators`, from P2P, TBD)
- at startup check if it matches the current, computed state hash
- if not, rollback to latest savepoint and sync from there

Olshansk replied:
> retrieve the latest valid statehash (from file, from SuperValidators, from P2P, TBD)

This is what "weak subjectivity" is. From [Vitalik's 2014 article](https://blog.ethereum.org/2014/11/25/proof-stake-learned-love-weak-subjectivity)

> Weak subjectivity is exactly the correct solution. It solves the long-range problems with proof of stake by relying on human-driven social information, but leaves to a consensus algorithm the role of increasing the speed of consensus from many weeks to twelve seconds and of allowing the use of highly complex rulesets and a large state. T

--- 

> if not, rollback to latest savepoint and sync from there
> at startup check if it matches the current, computed state hash

I think there are implementation nuances here, but I think you've got it.

--- 

I think it's worth adding (in the moonshot section) that we can potentially use Mina to verify the snapshot hash using only the latest state and the `genesis.json` file

https://minaprotocol.com/lightweight-blockchain

<img width="931" alt="Screenshot 2023-03-07 at 5 22 17 PM" src="https://user-images.githubusercontent.com/1892194/223594923-ffb42403-fafc-4c67-ac02-09e9e839db6a.png">

deblasis replied:
> I think there are implementation nuances here, but I think you've got it.

Music to my ears. 

I am going to read that article tonight. Thanks for sharing. 🙏

> I think it's worth adding (in the moonshot section) that we can potentially use Mina to verify the snapshot hash using only the latest state and the genesis.json file

Added

Olshansk commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+
+- Savepoints could be disseminated/retrieved using other networks like Tor/IPFS/etc., this would free-up bandwidth on the main network and could be used in conjunction with `FastSync` to speed up the process of bootstrapping a new node without overwhelming the `P2P` network with a lot of traffic that is not `Protocol`-related. This could be very important when the network reaches critical mass in ...
+
+  For example: a fresh node could be looking for the latest `Savepoint` signed by PNI available, download it from Tor, apply its state and resume the normal sync process from the other nodes from there.
```
`Tor` is great (and the right starting point) but long-term, try to also start thinking in terms of incentive: https://olshansky.medium.com/cryptocurrencies-its-all-about-incentive-77ac47a6adc4

- Is there another protocol that incentives sharing snapshots?
- Could/should we have another protocol actor for this?
- Is it part of Pocket or external?

For example, Gohkan and I were talking about this yesterday: https://olshansky.medium.com/cryptocurrencies-its-all-about-incentive-77ac47a6adc4

We got the idea of having SuperValidators (i.e. higher stake, higher penalty, limited in number) that could help with bootstrapping, and also make the formal 2/3 agreement more flexible. Could be a HotPOKT innovation ;)

deblasis replied:
Yeah, makes sense. Perhaps disseminating snapshots via Tor or other medium is something that I'd start doing as PNI (for example), then I'd look into offloading the responsibility to the protocol itself by creating the necessary incentives but first of all I'd like to see it working like a charm.

Since snapshots/rollback could be seen also as a "disaster recovery" tool as well, I'd advice to use something that's decoupled from Pocket, or at least that can be interacted with in other ways in the case of complete network halt.

I don't know why but I am thinking about Solana now :)

Olshansk replied:
Maybe add a "FiremanBoostrapper" or "Node911" as a moonshot idea :)

deblasis replied:
LOL, I'll leave that to you ;)

Olshansk replied:
Maybe even an IPFS hash that's stored directly in the block header

Olshansk commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+
+  For example: a fresh node could be looking for the latest `Savepoint` signed by PNI available, download it from Tor, apply its state and resume the normal sync process from the other nodes from there.
+
+- Building on top of the latter, when it comes to socially-coordinated rollbacks, the DAO/PNI could designate a `Savepoint` as the one that node-operators should use to rollback to by advertising it somewhere (this could be done in several ways, using alternative decentralized networks/protocols is preferable for increased resiliency: IPFS, Ethereum smart-contract interaction, etc.). This would a ...
```
Let's stick to social coordination for now. For "getting news", we might be able to leverage other protocols (Marlin?) that have incentives at the P2P layer, but I personally think this part is the hardest since it'll involve a lot of opinions as to how it should/shouldnt be done.

deblasis replied:
Agreed, just some random ideas really but anything can be iterated on I guess.

Let me know if I should amend something.

Olshansk replied:
Maybe add a "Moonshot ideas" section to it's clear that this is something you wanted on paper but don't plan to implement anytime soon.

Olshansk commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+
+## Random thoughts
+
+- I wonder if serialized, compressed and signed `Merkle Patricia Trie`s could be leveraged as a media for storing `Savepoint`s in a space-efficient and "blockchain-native" way 🤔.
```
Note that we're not using Etheruem's Merkle Patricia Trie, but a modified version of Libra's Jellyfish Merkle Tree implemented by Celestia.

The fact that we're using Trees, the root has a hash, and that hash is easily verifiable. If a node is synching from scratch, it's easy to verify the state transitions are correct. If a node is not synching from scratch, it needs to trust someone else. Mina protocol (which I don't really know) has a special property that if you have the genesis parameters (i.e. the json config) and ANY height, you can immediately (in O(1)) verify that it is correct. Might be worth looking into.

https://olshansky.substack.com/p/5p1r-ethereums-modified-merkle-patricia
https://olshansky.substack.com/p/5p1r-jellyfish-merkle-tree

deblasis replied:
Yeah, I was reading about Ethereum and I had a Freudian slip with Patricia, luckily Celestia is not jealous.... bad joke 😝

Noted.

Olshansk replied:
![7dnl7m](https://user-images.githubusercontent.com/1892194/223588017-22e4334e-79b6-41e3-adfc-bd02f0d63e8d.jpg)

Olshansk reviewed (changes requested):
I wanted to finish the first round of review with some thoughts/comments/ideas. You've identified a couple pieces that we can potentially start implementing right away, so let's continue the conversation. 

Keep throwing questions & ideas so we can iterate on it.

Two things:
1. I got a little tired toward the end so I apologize if the quality of the comments decreased.
2. In a couple places, I didn't go back to edit my previous comments, so some things may feel redundant.

deblasis commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+| Merkle Trees          | Merkle Trie backed by Key-Value Store | BadgerDB                          |
+
+Something worth mentioning specifically about `Merkle Trees` is the fact that we store separate trees for the individual `Actor` types (i.e. `App`, `Validator`, `Fisherman`, etc.), for the `Accounts`/`Pools` and for the data types such as `Transactions`, `Params` and `Flags`.
+
```
Yeah I have been looking into them also to understand if we are reinventing the wheel or not.

deblasis commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+
+### Savepoints
+
+A savepoint (also referred as checkpoints/snapshots depending on the implementation) is either the beginning of database transaction (or distributed transaction) or some sort of artifact that was created right after a successful commit happened that allows to recreate a perfect copy of the state at the time it was created.
```
TL;DR: yes

If the process that writes the files in the datadirs was stopped gracefully and there are no open handles on these files, a timestamped copy of these datadirs is what I would call `snapshot`.

Does this answer your question?

Olshansk replied:
It does answer my question, but I believe we are conflating `snapshot` and `savepoint`.

My original intention was:
* `snapshot` - A durable, shareable copy of the data directory.
* `savepoint` - an in-memory state that can be rolled back, updated or committed. It may or may not directly correspond to a `snapshot` that was stored on disk.

deblasis commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+
+### Rollbacks
+
+A rollback is the process of cleanly reverting the state of the node to a previous state that we have saved in a savepoint.
```
1. It captures both. If we want the ability of rolling back further in the past (block-height-wise) than the current non-committed ephemeral block, we need to be able to do that also for committed state as well.
2. Yes

Olshansk replied:
Noted. Resolving this but keeping the question here open: https://github.com/pokt-network/pocket/pull/533#discussion_r1128791630

deblasis commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+
+After having examined the `Persistence` and `Utility` modules, I have identified the following areas that we can consider as part of the MVP, these could be used as individual Github issues/deliverables potentially:
+
+- [**State changes invalidation and rollback triggering**] We need some sort of shared and thread-safe reference that is available for us across the whole call-stack that we can use in the event of a failure to flag that we need to abort whatever we are doing and rollback. This could be achieved via the use of the [context](https://pkg.go.dev/context) package.
```
Fair point. I was initially thinking about embedding the go `Context` into the `PersistenceContext` and in my research I bumped into this: 

![image](https://user-images.githubusercontent.com/29378614/222203512-3654c22e-f25c-4270-8ebe-0c72f34c9d24.png)

source: https://go.dev/blog/context-and-structs

Olshansk replied:
This is interesting and makes a ton of sense.

I completely agree that having `Context` be in a struct that also acts as a receiver is THE WRONG approach.

The very first thing we did though had a struct called `PocketContext` that contained golang `Context` inside of it, which got passed around everywhere (i.e. starting with `node.go`) as a first param. 

Don't have a strong opinion if we should just use `Context` everywhere though. It seems to be what everyone else is doing.

deblasis replied:
Devil's advocate: What if what everyone else is doing is wrong because they are all copying each other? (I should be spending more time looking at other projects, unfortunately I don't have good or bad examples, just assuming)

Or maybe sometimes a pattern makes sense but when applied it can get "lost in translation".

It makes me think about this:

![image](https://user-images.githubusercontent.com/29378614/223677621-bdeb987c-0e00-426f-96e1-3a069f6a3e2d.png)

Just kidding. 

The context is usually used to define the lifetime boundaries of what normally is an HTTP request.

In our case the "request" is effectively a "tentative world state transition application to the local state subject to byzantine consensus and unexpected errors".

Olshansk replied:
I had to tweet this: https://twitter.com/olshansky/status/1644090374242332672

deblasis commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+
+- [**State changes invalidation and rollback triggering**] We need some sort of shared and thread-safe reference that is available for us across the whole call-stack that we can use in the event of a failure to flag that we need to abort whatever we are doing and rollback. This could be achieved via the use of the [context](https://pkg.go.dev/context) package.
+
+- [**Ensure atomicity across data stores**] We need to make sure that we are using transactions correctly and that we are not accidentally committing state ahead of time in any of the data-stores.
```
Yeah, of course.

I would add that Postgres is a "historicized" view of the KV stores at different heights. 

That's why it's important to keep them in sync.

deblasis replied:
Yeah, I guess `indexed` works as well if we use a book as a metaphor: 
you see the height as a "chapter" you refer to from the index.

Since `height` has a temporal element in it, I conflated the two in `historicized`.

:)

deblasis commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+
+    > TODO: **Discuss what is appropriate to do here**
+
+- [**Savepoints**] The simplest version could be the database transaction that can simply be discarded, basically the uncommitted transaction sits in memory until it's flushed to storage and rolling back to a savepoint would be as simple as discarding the non-pristine version of the state in memory.
```
Just to make sure I understand, here's an example of what I envisage:

1. We have a committed height of 5 and we are at now, tentatively, at height 6, reaping the mempool, updating the trees, computing the root, computing the state hash
2. something (could be anything, including the fact that the state hash is not valid or non-consensus) goes wrong
3. the ephemeral changes are rolled back and we are back at the point **1** aka. the whole state is like it was at height 5 and we can try again

Am I missing something?

Olshansk replied:
Nope, we're on the same page.

- We use `postgres` for state change validation (can A send X POKT to B?)
- We use trees (k-v stores) to efficiently share the state everyone is agreeing on

- For postgers - rollback is easy because we keep reading/writing to an uncommited transaction and can just do `tx.rollback`.
- For KV stores, there is no `kv.rollback` without keeping track of all the modified keys

deblasis replied:
💯

deblasis commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+
+- [**Savepoints**] The simplest version could be the database transaction that can simply be discarded, basically the uncommitted transaction sits in memory until it's flushed to storage and rolling back to a savepoint would be as simple as discarding the non-pristine version of the state in memory.
+
+- [**Rollbacks**] Rolling back to a savepoint would mean not only that the state has been restored to the previous savepoint but also that the node has to go back into a state that allows it to proceed with its normal operation (i.e. all the modules should behave as if nothing has happened)
```
Yeah, potentially. 

If we did/do things well, the state should be completely a `persistence` concern and the other modules are more or less stateless consumers so we **shouldn't** have dangling/ephemeral stuff lying around in other modules. 

We might encounter something that needs refactoring along the way but that would be tech debt that we have to pay back I guess.

Olshansk replied:
> We might encounter something that needs refactoring along the way but that would be tech debt that we have to pay back I guess.

[A real software engineer always pays back his debts.](https://twitter.com/olshansky/status/1633274137576878080)

deblasis replied:
LOL I have so much tech debt accumulated in multiple countries that one day or another, with the rise of AI, it will become sentient and come after me. 😨

deblasis commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+
+- [**Rollbacks**] Rolling back to a savepoint would mean not only that the state has been restored to the previous savepoint but also that the node has to go back into a state that allows it to proceed with its normal operation (i.e. all the modules should behave as if nothing has happened)
+
+- [**Extensive testing**] We need to make sure that we have a good test coverage for all the above scenarios and that we can simulate failures in a controlled environment.
```
https://github.com/pokt-network/pocket/pull/533/commits/dbd741c4fddb2fb9a7728a8827fe5ad5852ca671

deblasis commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+Apart from internal failures that should resolve themselves automatically whenever possible, nodes might require a way to save their state and restore it later, not necessarily at the previous block height. This could be useful for a number of reasons:
+
+- To allow nodes to recover from an unforeseen crash (bugs)
+- To facilitate socially coordinated rollbacks of the chain to a specific block height in the case of a consensus failure/chain-halt
```
💯

deblasis commented:
@Olshansk I know you don't have a working laptop with you and I am not expecting an answer till you are back from Denver. Just an update:

I have addressed the comments and I am now going to rescope the related tickets to tackle the MVP scenario first. In the meantime I am working on #508 (keeping in mind the changes that will come with savepoints/rollbacks)

Perhaps I could create also tickets for the more advanced implementation but I would wait for this to be merged first.

Olshansk commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,146 @@
+
+We cannot really make the assumption, especially in this day and age, but in the simplest case, we could very much have a monolytic setup where the node is running on the same machine as the `PostgresSQL` database and the `BadgerDB` key-value stores but that would not change the fact that we are dealing with a distributed system since each one of these components is a separate process that could  ...
+
+Imagine a scenario in which the state is committed to the `PostgresSQL` database but one/some of the `BadgerDB` key-value stores fails due to storage issues. What would happen next? That's **non-deterministic**.
```
Perfect

Olshansk commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,147 @@
+
+- To allow nodes to recover from an unforeseen crash (bugs)
+- To facilitate socially coordinated rollbacks of the chain to a specific block height in the case of a consensus failure/chain-halt
+- To improve operator experience when managing and upgrading their fleet of nodes
```
```suggestion
- To improve operator experience when managing and upgrading their fleet of nodes
- Governance transactions that enable rolling back state subsets
```

Olshansk commented on persistence/docs/SAVEPOINTS_ROLLBACKS.md:
```diff
@@ -0,0 +1,147 @@
+- The action of creating a savepoint must be easy to perform (e.g. a CLI command and/or a flag being passed to the node binary when starting it)
+- The action of creating a savepoint must be as fast as possible
+- The operator should be informed with meaningful messages about the progress of the savepoint creation process (telemetry, logging and stdout come to mind)
+- It should be, as much as possible, compact
```
```suggestion
- It should be, as much as possible, compact (i.e. zipped) to reduce the size and cost of disseminating snapshots
- A reduction in the snapshot size should be prioritized over its compression speed since it is an infrequent event
```

Olshansk reviewed (changes requested):
> @Olshansk I know you don't have a working laptop with you and I am not expecting an answer till you are back from Denver. 

What a life, eh?  🤹 

---

> I have addressed the comments and I am now going to rescope the related tickets to tackle the MVP scenario first. In the meantime I am working on #508 (keeping in mind the changes that will come with savepoints/rollbacks)
> 
> Perhaps I could create also tickets for the more advanced implementation but I would wait for this to be merged first.

Discussed offline, but makes sense to me. Capturing some of the points

Between the
- (this) #493 
- (this extra) Adding `Context` everywhere
- (side quest) #508
- (untriaged) #361
- (legacy) #406

---

Checkout my responses to some of the comments. Hopefully both fun and educational.

Pretty sure you're deep in the implementation stage so I'm sure the ideas are flowing.

Olshansk reviewed (approved):
We're going to be figuring out implementation details along the way, but I think this is good to merge. Going to leave some of the comments unresolved because they're legendary :)

Just add a GITHUB_WIKI comment at the bottom before merging in, please!
